### ISSUE TITLE
Startrekker synthesized instruments should "overwrite" existing samples

### ISSUE THREAD

**Issue by sagamusix** [cmatsuoka/libxmp#94]
The code block for loading synthesized AM instruments is guarded by `if (mod->xxs[i].len == 0) {`.
Sample 1 of [fa.worse face.mod](https://www.exotica.org.uk/mediawiki/index.php?title=Special%3AModland&md=search&mod=fa.worse+face.mod&aut=&for=&col=&ins=) however has a non-zero length while being an AM instrument at the same time. I think the check for `am_synth && is_am_instrument(nt, i)` is enough in this case. In fact, I think the regular sample should always be loaded, and then replaced by the synthesized sample if requried, because otherwise the 64 bytes from the first sample will be added in front of the next sample, so all samples are shifted a bit.
In summary: Load the "normal" sample regardless of `is_am_instrument`, and then call `read_am_instrument` afterwards if required (do not forget to free the old sample in case this is not done automatically).

**Comment by sagamusix** [comment]
Another issue with the same file: The third and last instrument in the .nt file is truncated. Playing the file in Delix plays a square wave if you are lucky or something different if not, so that indicates that Delix reads from uninitialized memory. I would suggest to zero out the truncated parts of the instrument struct if it cannot be read completely, rather than not loading the instrument at all.

**Comment by sagamusix** [comment]
Another random find: The length of the final volume envelope segment should not be `256 / am.rs` but `am.sl / am.rs`.

**Comment by AliceLR** [comment]
> The code block for loading synthesized AM instruments is guarded by `if (mod->xxs[i].len == 0) {`.

Other modules with instruments where `am_synth && is_am_instrument(nt, i)` is true for an instrument with length != 0:
* `Broom/kingkeldon.mod` instrument 1: the AM instrument overwrites the bass drum sample when this check is removed.
* `WOTW/intro 12.mod` instrument 8: the sample is clearly very wrong here.

**Comment by AliceLR** [comment]
Per Asle, the .nt file for `Broom/kingkeldon.mod` is probably not authentic as it is not present in the demo, so it can be disregarded. In actual StarTrekker, it *does* override the bass drum until you free that AM instrument.

**Comment by AliceLR** [comment]
> Another random find: The length of the final volume envelope segment should not be `256 / am.rs` but `am.sl / am.rs`.

In general, the envelope timing math implemented by libxmp does not make any sense, and I'm trying to untangle that before I submit a PR for this.

```c
	if (am.a1l > am.l0) {
		a = am.a1l - am.l0;
		b = 256 - am.l0;
	} else {
		a = am.l0 - am.a1l;
		b = am.l0;
	}
	if (b == 0)
		b = 1;

	vol_env->data[2] = vol_env->data[0] + (256 * a) / (am.a1s * b);
	vol_env->data[3] = am.a1l / 4;
```
So... if these envelopes are purely linear like the StarTrekker docs imply, and if they update at 50Hz (every tick) like the StarTrekker docs outright state, `256` and `b` don't make any sense at all. The envelope stage duration should be a simple `ceil(abs(l1 - l0) / s1)`, which would give the correct release stage duration as reported.

flt_load.c cites StarTrekker documentation for this, but I have not found any such formula in any StarTrekker documentation so far.

edit: decreasing `L0` while leaving `A1S` at 1 in StarTrekker increases the length of the `A1` stage proportionally...

**Comment by AliceLR** [comment]
Both instruments (louder instrument in channel 1, quieter instrument doubled in channels 2 and 3) should end at about the same time, roughly row 16. libxmp gets this incorrect. Using the `ceil(abs(...)...)` calculation instead fixes this.

[mod.am env speed.zip](https://github.com/user-attachments/files/28789957/mod.am.env.speed.zip)

**Comment by AliceLR** [comment]
...and of course, once I get the envelope conversion actually fully working, I realize this feature may or may not override channel volume and may need extras. :/

edit: `Epsilon (ES)/frankie.mod` and `WOTW/into 12.mod` are mildly affected, whichever way this turns out. StarTrekker 1.2 definitely does not support volume effects with AM instruments, need to confirm other version(s).

edit: StarTrekker 1.3 also has instruments override channel volume. Specifically, channel volume seems to take effect for about 1/20th of a tick before AM instrument volume takes precedence again. Using `A0F` makes these "spikes" get progressively deeper, so it's not overwriting the channel volume (not that the distinction matters much, since playing a new non-AM instrument will set the volume anyway).
